src/dev/lupio/lupio_tty.cc

```diff
@@ -0,0 +1,176 @@
+#include "dev/lupio/lupio_tty.hh"
+
+#include "base/trace.hh"
+#include "debug/LupioTTY.hh"
+#include "dev/platform.hh"
+#include "mem/packet.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioTTY.hh"
+
+#define LUPIO_TTY_INVAL    0x80000000
+
+/* Same fields for CTRL and STAT registers */
+#define LUPIO_TTY_WBIT  (1 << 0)
+#define LUPIO_TTY_RBIT  (1 << 1)
+
+namespace gem5
+{
+
+/*
+ * IRQ management
+ */
+void
+LupioTTY::lupio_tty_update_irq()
+{
+    unsigned int irq;
+
+    irq = (writ_ie && writ_char != -1)
+        || (read_ie && read_char != -1);
+    if (irq) {
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, interrupting\n");
+        platform->postConsoleInt();
+    } else {
+//        platform->clearConsoleInt();
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, not interrupting\n");
+    }
+}
+
+void
+LupioTTY::dataAvailable()
+{
+      // unsure about these checks
+      if (read_ie && read_char == -1) {
+          DPRINTF(LupioTTY, "Data available, interrupting\n");
+          platform->postConsoleInt();
+     }
+}
+
+LupioTTY::LupioTTY(const Params &params)
+    : BasicPioDevice(params, params.pio_size), device(params.device),
+    byteOrder(params.byte_order),
+    platform(params.platform)
+{
+    // initialize internal registers
+    writ_char = read_char = -1;
```

Comment:
I wouldn't have multiple assignments on a single line. Also, for class member variables the initialization should be in the initializer list. OR, you can actually initialize in the header if the initial value is a constant.